### PR TITLE
feat(solana): Kamino Earn vault data layer

### LINKED ISSUE #2058: [ADD] Kamino Earn data layer — REST client, vault registry, positions/APY/PnL
## Summary

Chain-level foundation for Kamino Earn vaults in `@vultisig/core-chain`, as a new `chains/solana/kamino/` module (sibling of `chains/solana/staking/`). REST-only against `https://api.kamino.finance` (no API key), no UI — the UI lands in the client repos.

Parent: vultisig/vultisig-windows#4541 (cross-platform siblings: vultisig/vultisig-ios#5017, vultisig/vultisig-android#5485 — both shipped).

## Scope

**REST client**

- `GET /kvaults/vaults/{pubkey}` — vault state (shares mint, fees, `minDepositAmount`)
- `GET /kvaults/vaults/{pubkey}/metrics` — `apy30d`, TVL, share price
- `GET /kvaults/users/{pubkey}/positions` — all user positions
- `GET /kvaults/users/{user}/vaults/{vault}/pnl` (+ `/pnl/history`) — position PnL

Docs: https://kamino.com/docs/build/api-reference/introduction

**Vault registry** — the 3 launch vaults:

| Tier | Vault | Asset | Curator | Address |
|---|---|---|---|---|
| Conservative | Steakhouse USDC | USDC | Steakhouse Financial | `HDsayqAsDWy3QvANGqh2yNraqcD8Fnjgh73Mhb3WRS5E` |
| RWA yield | RWA USDC | USDC | RockawayX | `DWSXb18xZApz29vnQpgR2m6MynCT7PznaXt7Ut7M7KaP` |
| SOL yield | Allez SOL | SOL | Allez Labs | `A1so1bPD3W1TfeFwboDh8yfAAVaVtcdAYBYCjhg2mJQ` |

Curator and risk tier live in the registry because the API exposes neither. Registry addresses must be asserted valid Solana addresses at test time — a typo here is a money-critical constant. The RWA USDC vault is labeled one risk tier above the conservative vault (it lends against tokenized private credit), not as "T-bill backed".

**Typed amounts** — the deposit and withdraw endpoints take the same `amount` JSON field with *inverted units*: deposit is denominated in underlying tokens, withdraw in vault shares. Model these as two distinct branded types over `bigint` base units so passing one where the other belongs is a compile error, not a silently mis-sized transaction.

**Exact integer share/token conversion** — the share price arrives as a ~20-significant-digit decimal string. Parse it as numerator/10^scale and convert with integer arithmetic; never route it through `Number`/float. This is load-bearing: the API does not validate amounts, and a withdraw request above the user's share balance is rewritten to `u64::MAX` ("withdraw everything") — so rounding up by one base unit can turn a partial withdraw into a full exit.

**Service-boundary bounds checks** — strictly positive, within u64, sane decimal scale. The client is the only gate the API offers.

**Error envelope** — preserve both the HTTP status and Kamino's own error code: a retryable 503 arrives in the same JSON shape as a permanent 400.

## Acceptance criteria

- [ ] Vault state, metrics, positions, and PnL fetchable for the 3 launch vaults
- [ ] Registry extensible (new vault = registry entry only); addresses asserted valid in tests
- [ ] Token amounts and share amounts are not interchangeable at compile time
- [ ] Conversion tests cover share-base-unit boundary rounding (port the iOS `KaminoAmountTests` cases)
- [ ] Changeset bumps `@vultisig/core-chain`

## Reference implementation

iOS shipped this in vultisig/vultisig-ios#5053 (`KaminoAPI`, `KaminoService`, `KaminoModels`, `KaminoAmount`, `KaminoVaultRegistry`) with ~850 lines of tests worth porting.

### PR DESCRIPTION
**Pull request by Ehsan-saradar**
Closes #2058. First SDK slice of the Kamino Earn integration (vultisig/vultisig-windows#4541) — the read-side data layer only; transaction building, the attribution memo, and validation/decoding follow in #2059.

## What's in here

New module `packages/core/chain/chains/solana/kamino/`:

- **REST client** (`api.ts`) for the four read endpoints — vault state, metrics, user positions, PnL — over the shared `queryUrl`. Kamino's structured error body becomes a typed `KaminoServiceError` whose `reason` is a record union, keeping the transport status alongside the API's own error code: a retryable 503 arrives in the same body shape as a permanent 400, and `isRetryable` tells them apart.
- **Curated vault registry** (`registry.ts`) with the three launch vaults. Each entry pins the vault's immutable identity — token/shares mints, both decimal scales, the farm — plus curator and risk tier, which the API exposes nowhere. A transaction built by the API cannot be validated against values the same API supplied, so identity always comes from the registry, and hydration refuses a response that disagrees with it.
- **Typed amounts** (`amount.ts`): the deposit and withdraw endpoints take the same `amount` field with inverted units (tokens in, shares out), so token- and share-denominated amounts are distinct types over exact `bigint` base units — passing one where the other belongs is a compile error.
- **Exact rate math** (`rate.ts`, `decimal.ts`): the ~20-significant-digit share rate is held as `numerator/10^scale` and every conversion is integer arithmetic truncating toward zero. This is load-bearing: the API does not validate amounts and rewrites a withdraw at/above the share balance to `u64::MAX` ("withdraw everything"), so a division that rounded up by one base unit could turn a partial withdraw into a full exit. Rounded-up variants exist solely for rendering minimums and are documented as never transaction-sizing.
- **Vault hydration** (`vaultInfo.ts`): merges registry identity with live state/metrics and derives the *effective* minimums the on-chain program actually accepts — the published `minDepositAmount` is refused at face value by the program (margin added), and `minWithdrawAmount` is token-denominated despite its name, with the real floor sitting above it (3× multiple, converted to shares rounded up).
- **Position parsing** (`position.ts`): truncates the API's over-precise share balances (up to 14 decimals for a 6-decimal mint) to the mint scale, computes a spendable maximum strictly below the reported balance (−1 base unit when exactly representable, since a request *at* the balance triggers the withdraw-everything sentinel), and cross-checks the staked/unstaked/total split at full reported precision.

Bounds are enforced at the service boundary (strictly positive, within `u64`, sane decimal scale) because the client is the only gate the API offers.

## Not in here (by design)

`POST /ktx/kvault/deposit|withdraw` builders, memo injection, blockhash refresh, and the transaction validator/decoder — all scoped to #2059.

## Testing

- 57 colocated unit tests: strict-format decimal refusals, exact rate parsing past double precision, conversion truncation/round-trip invariants (a sized withdraw never exceeds what was asked), the differing-decimals vault (token 9 / shares 6), registry address validity (every pinned address must parse as a Solana `PublicKey`), effective-minimum derivations, identity-mismatch refusals, spendable-balance edge cases, and the error envelope.
- Full core suite passes (2,838 tests), plus lint, prettier, knip, the shared-exports check, and `yarn changeset:check-sdk-bundle`.
- Changeset bumps `@vultisig/core-chain` and `@vultisig/sdk` (minor).

## Summary by CodeRabbit

- **New Features**
  - Added Solana Kamino Earn support with live vault metrics, positions, PnL, and curated vault information.
  - Added precise token/share conversions and high-precision rate handling.
  - Added vault validation, deposit and withdrawal minimum calculations, risk metadata, and safe spendable balance handling.
  - Added structured service errors with retryability indicators for temporary failures.
- **Tests**
  - Added comprehensive coverage for amounts, rates, decimals, API behavior, vault validation, and position parsing.

### CODE REVIEW

File: packages/core/chain/chains/solana/kamino/amount.ts [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  rate.scale <= kaminoMaxDecimals * 4 &&
+  decimalScales.every(decimals => decimals >= 0 && decimals <= kaminoMaxDecimals)
+
+const divide = (numerator: bigint, denominator: bigint, roundUp: boolean): bigint =>
+  roundUp ? (numerator + denominator - 1n) / denominator : numerator / denominator
+
+const parseBaseUnitString = (raw: string): bigint | undefined => {
+  if (!/^-?\d+$/.test(raw)) return undefined
+  return BigInt(raw)
+}
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply packages/core/chain/chains/solana/kamino/amount.ts]: Fixed in 3eb76a1cc — the regex is now `/^\d+$/` and signed input (including `-0`) returns undefined, with regression tests.

File: packages/core/chain/chains/solana/kamino/api.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ */
+export const fetchKaminoUserPositions = (owner: string): Promise<KaminoUserPositionResponse[]> =>
+  kaminoGet(`/kvaults/users/${owner}/positions`, positionsReadTimeoutMs)
+
+/**
+ * Fetches lifetime PnL for one position
+ * (`GET /kvaults/users/{owner}/vaults/{vault}/pnl`).
+ */
+export const fetchKaminoPnl = ({ owner, vault }: { owner: string; vault: string }): Promise<KaminoPnlResponse> =>
+  kaminoGet(`/kvaults/users/${owner}/vaults/${vault}/pnl`)
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply packages/core/chain/chains/solana/kamino/api.ts]: Fixed in 3eb76a1cc — all four fetchers now encode their path segments, and a test asserts a hostile value ends up percent-encoded instead of retargeting the endpoint.

File: packages/core/chain/chains/solana/kamino/position.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    staked,
+    unstaked,
+    total,
+    spendable,
+    accountsForItsTotal: reportedSum ? kaminoRateEquals(reportedSum, position.totalShares) : false,
+    isPlausible:
+      staked.baseUnits >= 0n &&
+      unstaked.baseUnits >= 0n &&
+      staked.baseUnits <= total.baseUnits &&
+      unstaked.baseUnits <= total.baseUnits,
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply packages/core/chain/chains/solana/kamino/position.ts]: Fixed in 3eb76a1cc — isPlausible now compares staked+unstaked against the total at the API's reported precision via a new compareKaminoRates helper. Added the 2+2 vs 3 regression plus a case where the excess hides past the mint scale (1.0000015 vs 1.0000011).

File: packages/core/chain/chains/solana/kamino/vaultInfo.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    name: state.name,
+    minDeposit,
+    minWithdraw,
+    lookupTable: state.vaultLookupTable,
+    apy30d,
+    tokensPerShare,
+    tokenPriceUsd,
+    // Advisory, so an unreadable value drops the liquidity notice rather than
+    // failing the whole hydration and blocking deposits too.
+    tokensAvailable: kaminoTokenAmountFromDecimalString(metrics.tokensAvailable, descriptor.tokenDecimals),
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply packages/core/chain/chains/solana/kamino/vaultInfo.ts]: This one can't happen: kaminoTokenAmountFromDecimalString routes through scaleKaminoRate, which refuses negative numerators, so tokensAvailable: '-1' already comes back undefined. Added a regression test in 3eb76a1cc pinning that behavior.

**Review by gomesalexandre — APPROVED** [review]
looks good.

receipts on `3eb76a1cc1eaae8566194d510c5e5900f8b5f396`:
- exact-head focused core receipt: `./node_modules/.bin/vitest run --config .config/vitest.core.config.ts packages/core/chain/chains/solana/kamino/{amount,api,registry,rate,vaultInfo,position,decimal}.test.ts` -> 7 files passed, 63 tests passed
- exact-head manual read of the hardened guards in `packages/core/chain/chains/solana/kamino/api.ts`, `vaultInfo.ts`, and `position.ts` looks sound. the current-head follow-up now path-encodes request segments and keeps the plausibility / spendable-share checks fail-closed.
- `yarn build:shared` is red here, but it reproduces unchanged on `origin/main` with the pre-existing cardano resolver error in `packages/core/mpc/keysign/signingInputs/resolvers/cardano.ts:72` (`auxiliaryData` not in `ISigningInput`), so i did not treat that as a regression from this pr.

nice hardening pass.

**Review by gomesalexandre — APPROVED** [review]
Read-side data layer only (no tx building yet, deferred to #2059), 57 unit tests covering rate math/truncation/identity-mismatch refusals, full core suite + lint/knip/changeset-bundle-check clean, changeset present. Solid. Approving.
